packages/compiler-vapor/src/generators/expression.ts

```diff
@@ -209,3 +211,347 @@ function canPrefix(name: string) {
+  const seenIdentifier = new Set<string>()
+
+  const registerVariable = (
+    name: string,
+    exp: SimpleExpressionNode,
+    isIdentifier: boolean,
+  ) => {
+    if (isIdentifier) seenIdentifier.add(name)
+    seenVariable[name] = (seenVariable[name] || 0) + 1
+    variableToExpMap.set(
+      name,
+      (variableToExpMap.get(name) || new Set()).add(exp),
+    )
+    expToVariableMap.set(exp, (expToVariableMap.get(exp) || []).concat(name))
+  }
+
+  for (const exp of expressions) {
+    if (!exp.ast) {
+      exp.ast === null && registerVariable(exp.content, exp, true)
+      continue
+    }
+
+    walk(exp.ast, {
+      enter(currentNode: Node) {
+        if (currentNode.type === 'MemberExpression') {
+          const memberExp = extractMemberExpression(
+            currentNode,
+            (name: string) => {
+              registerVariable(name, exp, true)
+            },
+          )
+          registerVariable(memberExp, exp, false)
+          return this.skip()
+        }
+
+        if (currentNode.type === 'Identifier') {
+          registerVariable(currentNode.name, exp, true)
+        }
+      },
+    })
+  }
+
+  return { seenVariable, seenIdentifier, variableToExpMap, expToVariableMap }
+}
+
+function processRepeatedVariables(
+  context: CodegenContext,
+  seenVariable: Record<string, number>,
+  variableToExpMap: Map<string, Set<SimpleExpressionNode>>,
+  expToVariableMap: Map<SimpleExpressionNode, string[]>,
+  seenIdentifier: Set<string>,
+): DeclarationValue[] {
+  const declarations: DeclarationValue[] = []
+  for (const [name, exps] of variableToExpMap) {
+    if (seenVariable[name] > 1 && exps.size > 0) {
+      const isIdentifier = seenIdentifier.has(name)
+      const varName = isIdentifier ? name : genVarName(name)
+
+      // replaces all non-identifiers with the new name. if node content
+      // includes only one member expression, it will become an identifier,
+      // e.g., foo[baz] -> foo_baz.
+      // for identifiers, we don't need to replace the content - they will be
+      // replaced during context.withId(..., ids)
+      const replaceRE = new RegExp(escapeRegExp(name), 'g')
+      exps.forEach(node => {
+        if (node.ast) {
+          node.content = node.content.replace(replaceRE, varName)
+          // re-parse the expression
+          node.ast = parseExp(context, node.content)
+        }
+      })
+
+      if (
+        !declarations.some(d => d.name === varName) &&
+        (!isIdentifier || shouldDeclareVariable(name, expToVariableMap, exps))
+      ) {
+        declarations.push({
+          name: varName,
+          isIdentifier,
+          value: extend(
+            { ast: isIdentifier ? null : parseExp(context, name) },
+            createSimpleExpression(name),
+          ),
+          rawName: name,
+          exps,
+          seenCount: seenVariable[name],
+        })
+      }
+    }
+  }
+
+  return declarations
+}
+
+function shouldDeclareVariable(
```

Comment:
Probably not a bottleneck but this function could use some optimizations to avoid too many iteration method calls